### PR TITLE
crimson/net: prevent reusing the sent messages 

### PR DESCRIPTION
**Pull request by cyx1231st**

### CODE REVIEW

File: src/test/crimson/test_messenger.cc [hunk 1]
```diff
@@ -51,7 +50,7 @@ static seastar::future<> test_echo(unsigned rounds,
           logger().info("server got {}", *m);
         }
         // reply with a pong
-        return c->send(msg_pong);
+        return c->send(MessageRef{new MPing(), false});
```
Comment (tchaikov) [inline src/test/crimson/test_messenger.cc]: might want to use `make_message()`

File: src/test/crimson/test_messenger.cc [hunk 2]
```diff
@@ -230,7 +228,7 @@ static seastar::future<> test_echo(unsigned rounds,
                                 seastar::stop_iteration::no);
                             });
                         } else {
-                          return conn->send(msg_ping)
+                          return conn->send(MessageRef{new MPing(), false})
```
Comment (tchaikov) [inline src/test/crimson/test_messenger.cc]: ditto.

File: src/crimson/net/Protocol.cc [hunk 3]
```diff
@@ -101,19 +101,19 @@ seastar::future<stop_t> Protocol::do_write_dispatch_sweep()
     size_t num_msgs = conn.out_q.size();
     // we must have something to write...
     ceph_assert(num_msgs || need_keepalive || keepalive_ack.has_value());
-    Message* msg_ptr = nullptr;
+    MessageRef front_msg = nullptr;
```
Comment (tchaikov) [inline src/crimson/net/Protocol.cc]: nit, s/= nullptr//.

File: src/test/crimson/test_messenger.cc (hunk 1 above)
Comment (cyx1231st) [reply src/test/crimson/test_messenger.cc]: done.

File: src/test/crimson/test_messenger.cc (hunk 2 above)
Comment (cyx1231st) [reply src/test/crimson/test_messenger.cc]: done.

File: src/crimson/net/Protocol.cc (hunk 3 above)
Comment (cyx1231st) [reply src/crimson/net/Protocol.cc]: done.

**Comment by cyx1231st** [comment]
> could you elaborate a little bit in the commit message that why we want to avoid reusing the sent messages?

see https://github.com/ceph/ceph/pull/28890/commits/2d6e77049259daf6521849ab8404b975965fa488

**Pull request merged**